Add OpenUserCSS support

DecentM opened the pull request:
This PR adds a handshake between the site and Stylus.
Issue: #326 

Before merging: I haven't been able to get the content script to run in Firefox, but it does work in Chromium.

tophf commented on content/openusercss-integration.js:
```diff
@@ -0,0 +1,58 @@
+        'key':       event.data.key,
+        'extension': {
+          'name':         manifest.name,
+          'version':      manifest.version,
```
There shouldn't be any need for the extension's manifest, it's irrelevant to the functionality.

DecentM replied:
I don't think it's irrelevant, because the site shows these values: 
In the install button: ![screenshot](https://i.imgur.com/0IEJCNW.png)
On the help page: ![screenshot](https://i.imgur.com/Pi8xdlP.png)

tophf replied:
The users don't like being tracked. There's no need to check for the version. As you know, it's always preferable to test for specific features like you already do.

DecentM replied:
I don't see how showing the version number on the help page could be seen as tracking, but yes, there's not a critical need for it.
I made it optional for extensions in the client javscript and removed the version number from the response object.

tophf replied:
Users are that weird. Not sure why you're still using manifest though. The extension name shouldn't serve any useful purpose - the capabilities is the only meaningful object there.

DecentM replied:
The site doesn't assume the extension name, so it has to get it from somewhere.
The install style button is displaying "Install theme with {{extension.name}}".

tophf replied:
Ah, makes sense then.

tophf commented on manifest.json:
```diff
@@ -59,6 +59,12 @@
+      "matches": ["https://openusercss.org/*", "https://openusercss.com/*"],
+      "run_at": "document_start",
+      "all_frames": false,
+      "js": ["content/openusercss-integration.js"]
```
Other content scripts use a different naming scheme so this one should be named as `install-hook-openusercss.js`.

DecentM replied:
While the filename doesn't hold much meaning in terms of function, I named it like this because there isn't any install hook in it yet, only the handshake. If you insist, I don't mind renaming it at all.

tophf replied:
Well, the primary use case will be installation of styles so it's better to have the name follow the scheme.

DecentM replied:
That's a good point, it will have that down the line, I renamed it :+1:

tophf commented on content/openusercss-integration.js:
```diff
@@ -0,0 +1,58 @@
+  // Tell the page that we exist and that it should send the handshake
+  window.postMessage({
+    'type': 'ouc-begin-handshake'
+  }, '*');
```
I think you should specify `location.origin` here.

DecentM replied:
I don't see the way that would be useful or necessary. Can you elaborate what you mean?

tophf replied:
Just for the sake of completeness because this is how postMessage is supposed to be used. Theoretically though, another extension could open this tab and send a message from another tab :-)

tophf replied:
Actually, I would prefer a dedicated custom DOM event. Why are you using postMessage? It's a busy one.

DecentM replied:
I pushed an origin whitelist, which covers for this and the event origin.
I'm using postMessage because when I use `CustomEvent`s, they don't trigger the listener in the content script. There's probably some sort of scoping, but checking for the `type` property filters out everything we don't need.

tophf replied:
Stylish/Stylus have always been using custom DOM events on USO so it might be better if you investigate why it doesn't trigger here.

DecentM replied:
After looking into it, I think it's because the content security policy the site uses (https://www.w3.org/TR/CSP3/#should-block-inline).
I think it wouldn't be a good idea allow inline event handlers from just anywhere. I believe sendMessage uses the same message bus (on `window`) under the hood, so I'm pretty sure there's no significant drawback in doing it the current way.

tophf replied:
Stylus doesn't depend on inline event listeners because we bind them in js as you can see in [install-hook-userstyles.js](https://github.com/openstyles/stylus/blob/master/content/install-hook-userstyles.js#L14). BTW, I would bind them on `window` object so that they don't pollute the devtools elements inspector.

DecentM replied:
Sorry, I think I used the wrong terminology here. What I wanted to mean is that listeners not added by a script that exists in the HTML head are not allowed to listen for custom events.  

All event listeners are bound to `window`.
In this PR: https://github.com/openstyles/stylus/pull/328/commits/6657ab53647e86dab23791aabc02c358ce941015#diff-057c3ffbe917f05ee2f42ce8d36884a9R18
On the site: [OpenUserCSS/openusercss.org/blob/extension-integration/src/client/js/browser.js#L93](https://github.com/OpenUserCSS/openusercss.org/blob/extension-integration/src/client/js/browser.js#L93)

tophf replied:
1. Scope of custom DOM events propagation isn't limited by CSP as far as I know, at least I don't see anything about that in the specification.
2. `message` event is always posted to the `window`, but I'm talking about custom DOM events

tophf replied:
Here's an example:

![pic](https://puu.sh/yVrPI/4e4e8f44e2.png)

1. the first line is executed in the extension's content script (using the toolbar selector)
2. the second line is executed here in github page context (using the toolbar selector)

As you can see the message is successfully delivered. The only complication is FF needs [`cloneInto`](https://github.com/openstyles/stylus/blob/master/content/install-hook-userstyles.js#L94).

tophf replied:
Well, come to think of it, we can use `message` on `window` since it works already.

DecentM replied:
Awesome!
Implementing CustomEvents down the line could be useful though, I think it's definitely solvable.

tophf commented on content/openusercss-integration.js:
```diff
@@ -0,0 +1,58 @@
+
+  // Wait for the handshake
+  window.addEventListener('message', event => {
+    if (event.data && event.data.type === 'ouc-handshake-question') {
```
You should check `event.origin` too.

DecentM replied:
The manifest already defines that the content script can only run on the site itself. Are there other cases in which an event could be received?

tophf replied:
See my comment above.

tophf commented:
So when do you think it will be able to actually install styles (at least)? I really don't want to add a dummy code :-)

DecentM commented:
I'm working on https://github.com/OpenUserCSS/openusercss.org/issues/16 as we speak. After that, I'll implement an event based theme installation.  
Usercss formatted themes can already be installed. The next release of the site will also promote Stylus in the help page if this handshake isn't completed correctly. That's why I made the initial PR short, so that a warning can be shown if the client is certain that the user won't be able to install themes.  

What I need help with though is an install function, because I don't know the inner workings of the codebase. If you could, please create a function that has an API that roughly looks like this:
```js
window.addEventListener('...', () => {
    installTheme({
        'name': 'My theme',
        'code': '/* ===userstyle=== ...'
    })
})
```
I've also [added capabilities](https://github.com/OpenUserCSS/openusercss.org/blob/master/src/client/js/browser.js#L76) to the site client that signals that Stylus exposes methods to query installed themes, reinstall themes, etc.

tophf commented:
Assuming `json` is the style object that has `name` string and `code` string,

```js
chrome.runtime.sendMessage({
  method: 'saveUsercss',
  reason: 'install',
  name: json.name,
  sourceCode: json.code,
});
```

tophf commented:
I guess we'll need to save the id of OUCSS style to use it when checking for updatability of the style currently opened on the site, currently there's no support for this.

DecentM commented:
Should be relatively easy, the API exposes the last update time.
The query would look like this:
```js
query {
  theme(id: "${id}") {
    lastUpdate
  }
}
```

Another approach would be to simply XHR https://api.openusercss.org/theme/${id} - it returns a full usercss

tophf commented:
It might be easier if the site id is passed along with that json on installation so we can save it in our style object. OTOH maybe you should just check `@namespace` and `@name`, I just don't know.

@Mottie, @eight04, ping.

Mottie commented:
IIRC, using `@namespace` and `@name` is the way it's checked internally. Then the `@version` is compared to check for an update.

eight04 commented:
> save the id of OUCSS style

It's not a good idea. Usercss should work without openusercss.org. If the site needs the information of installed style, it can simply ask the extension (post a message and get the response). We just need to expose the API in `install-hook-openusercss.js`.

----

I think we should implement some kinds of `.meta.css` (in other PR) to check update for usercss. A `.meta.css` file is basically a text file containing the metadata block, with some basic information like name, namespace, and version.

For example, the user has installed a style from `https://api.openusercss.org/theme/5a50a07bd4c8c00b008370f1.user.css`. To check update, Stylus first requests `https://api.openusercss.org/theme/5a50a07bd4c8c00b008370f1.meta.css` for metadata. If an update is available (version is changed), then it could download the whole usercss from `https://api.openusercss.org/theme/5a50a07bd4c8c00b008370f1.user.css`.

As a result, the hoster (e.g. openusercss.org) can serve a much smaller file to save bandwidth.

DecentM commented:
I added functionality to install usercss through window events.
I also pushed an update to the site that can use this, the only thing that remains is a way to query if a theme is installed.

@eight04 This is already possible by querying the public API, see https://github.com/openstyles/stylus/issues/326#issuecomment-355736958  

@tophf Can you help me out with a function that returns an array of installed styles please?

tophf commented:
>function that returns an array of installed styles

That sounds like an intrusion on privacy. How about I only expose a way to check if the currently displayed style is installed?

DecentM commented:
Yep, that's exactly what I meant, like the one you wrote above to be inserted into the extension.  
Do you have an idea for what the key to base the query on should be?

tophf commented:
Stylus considers usercss equal if both its `@name` and `@namespaces` are equal to the other style's so the site only needs to provide those to the content script. I've just added the method so you can use it in the content script:

```js
chrome.runtime.sendMessage({
    method: 'findUsercss',
    name: 'foo',
    namespace: 'bar',
  }, style => {
  console.log(style);
});
```

The result is a style object, may be undefined/null if no match was found.

Mottie commented:
> Stylus considers usercss equal if both its `@name` and `@namespace` are equal

It would be nice to also include the version, that way you can tell if the usercss is "really" equal, or needs updating.

tophf commented:
The version is returned in the style object so the content script can of course check the version as well as variables and whatever is relevant.

DecentM commented:
I think it's ready for merging now.
I also added event listeners for some API requests, let me know if I missed anything!

DecentM commented:
@tophf Can you review please?

tophf commented on background/openusercss-api.js:
```diff
@@ -0,0 +1,106 @@
+// end:nanographql
+
+const api = 'https://api.openusercss.org';
+const doQuery = ({id}, queryString) => new Promise((resolve, reject) => {
```
No need for a new Promise, you can chain on `fetch` and return it. To reject simply return `Promise.reject(whatever)`

tophf commented on background/openusercss-api.js:
```diff
@@ -0,0 +1,106 @@
+  })
+  .then(res => {
+    res.json()
+    .then(response => {
```
With the above note implemented, this `.then()` can be moved up a level and the entire chain flattened.

tophf commented on content/install-hook-openusercss.js:
```diff
@@ -0,0 +1,159 @@
+'use strict';
+
+const manifest = chrome.runtime.getManifest();
```
Wrap the entire code in IIFE because content scripts may be reinjected on update or on manual reload and there are problems with global const/let at least in some browser versions. See install-hook-userstyles.js on how to check for the orphaned state.

tophf replied:
install-hook-usercss.js might be a better example of orphanCheck

tophf commented on content/install-hook-openusercss.js:
```diff
@@ -0,0 +1,159 @@
+      && allowedOrigins.includes(event.origin)
+    ) {
+      chrome.runtime.sendMessage({
+        'method':    'findUsercss',
```
We don't quote key names.

tophf commented on content/install-hook-openusercss.js:
```diff
@@ -0,0 +1,159 @@
+          sendInstalledCallback({
+            'installed': true,
+            'enabled':   style.enabled,
+            'name':      event.data.name,
```
Make it DRY by using shorthand initializers for `name` and `namespace`. Or maybe make one object for the message, change the relevant properties, then send it using just one sendInstalledCallback call.

tophf commented on content/install-hook-openusercss.js:
```diff
@@ -0,0 +1,159 @@
+      && allowedOrigins.includes(event.origin)
+    ) {
+      chrome.runtime.sendMessage({
+        'method':     'saveUsercss',
```
We don't align key:value properties.

DecentM commented:
Thanks! I fixed the issues you mentioned. What do you think?

tophf commented:
I don't see orphanCheck logic anywhere.

DecentM commented:
On [line 156](https://github.com/openstyles/stylus/pull/328/files#diff-057c3ffbe917f05ee2f42ce8d36884a9R156), all event listeners are removed. There are no other permanent effects of the content script, since everything is wrapped in a function.  
An orphan check for these three event listeners would be more complex and error prone.

tophf commented:
Uhm, removeEventListener won't remove the old orphaned instance of the function because that function **reference** is different from the reinjected one, which is why orphanCheck logic is needed, and it shouldn't introduce any problems since all it does is notifying the orphaned instance it's been superseded.

DecentM commented:
Thanks for clarifying that! I think I understand what you mean.  
I updated the branch with a check based on the one in `install-hook-usercss.js`.

tophf commented on content/install-hook-openusercss.js:
```diff
@@ -0,0 +1,178 @@
+  ];
+
+  const sendPostMessage = message => {
+    allowedOrigins.forEach(origin => {
```
`forEach` is not needed anymore.

DecentM replied:
The reason I left it in is that `window.postMessage(message, origin);` needs an origin. If the forEach was removed, `origin` would become undefined.
What do you think would be the ideal solution to this?

tophf replied:
Simply use location.origin. Current code doesn't make sense.

DecentM replied:
Oh, I see, thanks! This is what happens when I keep overthinking things. 🚀

tophf commented:
Regarding the UX, currently it's weird for: two tabs are opened just to install a style. I think no tabs should open at all, the style should be installed, the page should reflect the new state by renaming its buttons and whatnot.

Also, the branch is not rebased onto the current master.

DecentM commented:
The flow you're talking about happens when the extension isn't detected on the page. In this case the site falls back to the traditional `usercss` style install, in which it opens up a page with the theme source code.  
I think it would be a good idea to introduce a new property like `@silent` to the usercss spec that will skip opening the style editor after installation. This would get rid of the second opening tab. After this, the source tab can be closed by the site that opened it, resulting in a flash, that's arguably better than two open tabs.
This is outside the scope of this PR, but I agree that the UX could be better in this scenario.

Fixed the rebase, also fixed a merge conflict :tada:

DecentM commented:
Oh also, changing the button text is already implemented in [master](https://github.com/OpenUserCSS/openusercss.org/tree/master), it's just not live yet.

tophf commented:
>I think it would be a good idea to introduce a new property like @silent to the usercss spec

Why? My point is, I don't see the need to open the style code so it should be always silent.

DecentM commented:
What I mean is that this behaviour exists on every usercss page, not just OUC.  
For example, if you open https://raw.githubusercontent.com/RaitaroH/YouTube-DeepDark/master/YouTubeDeepDarkMaterial.user.css, the same thing happens with the new tab being opened.  

I made a short clip of how it will work once the PR is merged and the site is updated to include the event driven installation method. https://youtu.be/sgbF3GcjK6A  
The help page already has validations and notice that appears when the handshake isn't completed.

tophf commented:
Installing from a specialized site should be more seamless. Both userstyles.org or freestyler.ws don't open any additional tabs and I think OUCSS should follow the suit.

eight04 commented:
@tophf 

The source code page is not closed after the installation.

https://github.com/openstyles/stylus/blob/8821c2f23f380d98262637f7c2df599275099d03/content/install-hook-usercss.js#L53-L59

Something went wrong here:

https://github.com/openstyles/stylus/commit/aedb02bbb19356fe069996c3d21577a88ed5cf20#diff-e0e82cc2f07995a5e5771a1cf2a8a8c1R190

https://github.com/openstyles/stylus/commit/b762f48135a9ad80bfb0a76b87b0ccd098e6f02e#diff-e0e82cc2f07995a5e5771a1cf2a8a8c1

tophf commented:
@eight04, it works as intended: we should not close a tab explicitly opened by the user. Currently, OUCSS site detects Stylus and shows "Install theme with Stylus" (BTW, the button text flickers during page load), but it opens the style tab anyway, which is wrong.

DecentM commented:
The live version of the site doesn't have support for event based theme installation right now, so instead, it just inserts the extension name and displays the feature list on the help page, but otherwise works as if no extension was detected in terms of installing themes.  

Once I finish the currently cooking version, it will work like it does in the video above.

tophf commented:
Just to confirm, you're sure it's server-side only and this PR doesn't need to be modified?

DecentM commented:
It's client-side, but I'm pretty sure Stylus won't need to be modified. The video I uploaded was recorded on the latest commit in master, which is far beyond what's on the live site.

DecentM commented:
Thank you! :tada:
